quark/subproject.py

```diff
@@ -319,6 +327,33 @@ def actualUpdate():
                     fork(['git', 'checkout', 'FETCH_HEAD', '--'])
                 else:
                     fork(['git', 'fetch'])
+                    # If we want to go on a branch, try to find a local branch that tracks it
+                    # and use it (possibly with a fast-forward)
+                    if self.ref_type == 'branch':
+                        # Resolve the remote ref
+                        remote_fullref = self.symbolic_full_name(self.ref)
+                        # Get a sensible local branch name to try
+                        local_ref = remote_fullref.split('/origin/', 1)[1]
+                        # Check if it is actually tracking our target
+                        try:
+                            local_fulltrackref = self.symbolic_full_name(local_ref + "@{u}")
+                        except CalledProcessError:
+                            # It's fine if it fails - we may not have a local-tracking branch,
+                            # so git checkout will do the right thing here
+                            local_fulltrackref = remote_fullref
+
+                        if remote_fullref == local_fulltrackref:
+                            try:
+                                # Checkout and fast-forward
+                                fork(['git', 'checkout', local_ref, '--'])
+                                fork(['git', 'merge', '--ff-only', self.ref, '--'])
+                                # Final sanity check
+                                if log_check_output(['git', 'rev-parse', self.ref]) != log_check_output(['git', 'rev-parse', local_ref]):
+                                    logger.warning("Warning: your local branch is ahead of required remote branch!")
+                                return
+                            except CalledProcessError:
+                                logger.warning("Couldn't fast-forward local branch, fallback to detached head mode...")
+                    # General case: plain checkout of the origin ref (going in detached HEAD)
                     fork(['git', 'checkout', self.ref, '--'])
```

Comment:
Wouldn't this create a detached HEAD pointing to remote even when we want to use the local tracking branch?

Reply:
Nope, there's a `return` at line 353 that avoids this `git checkout` if everything goes well in the code path of the local branch.